augur/utils.py

```diff
@@ -143,66 +144,272 @@ def default(self, obj):
-                            fname = feat.qualifiers["locus_tag"][0]
-                        else:
-                            fname = None
-                    else:
-                        if "gene" in feat.qualifiers:
-                            fname = feat.qualifiers["gene"][0]
-                        elif "gene_name" in feat.qualifiers:
-                            fname = feat.qualifiers["gene_name"][0]
-                        else:
-                            fname = feat.qualifiers["locus_tag"][0]
-                    if feat.type == "source":
-                        fname = "nuc"
-
-                    if fname:
-                        features[fname] = feat
-
-            if feature_names is not None:
-                for fe in feature_names:
-                    if fe not in features:
-                        print("Couldn't find gene {} in GFF or GenBank file".format(fe))
-
+        return _read_gff(reference, feature_names)
     else:
-        from Bio import SeqIO
-        for feat in SeqIO.read(reference, 'genbank').features:
-            if feat.type=='CDS':
-                if "locus_tag" in feat.qualifiers:
-                    fname = feat.qualifiers["locus_tag"][0]
-                    if feature_names is None or fname in feature_names:
-                        features[fname] = feat
-                elif "gene" in feat.qualifiers:
+        return _read_genbank(reference, feature_names)
+
+def _read_nuc_annotation_from_gff(record, reference):
+    """
+    Looks for the 'record' & 'source' GFF type as well as the ##sequence-region
+    pragma. Note that 'source' isn't really a GFF feature type, but is used
+    widely in the Nextstrain ecosystem. If there are multiple we check that the
+    coordinates agree.
+    
+    Parameters
+    ----------
+    record : <class 'Bio.SeqRecord.SeqRecord'> reference: string
+
+    Returns
+    -------
+    <class 'Bio.SeqFeature.SeqFeature'>
+
+    Raises
+    ------
+    AugurError
+        If no information on the genome / seqid length is available or if the
+        information is contradictory
+    """
+    nuc = {}
+    # Attempt to parse the sequence-region pragma to learn the genome
+    # length (in the absence of record/source we'll use this for 'nuc')
+    sequence_regions = record.annotations.get('sequence-region', [])
+    if len(sequence_regions)>1:
+        raise AugurError(f"Reference {reference!r} contains multiple ##sequence-region pragma lines. Augur can only handle GFF files with a single one.")
+    elif sequence_regions:
+        from Bio.SeqFeature import SeqFeature, FeatureLocation
+        (name, start, stop) = sequence_regions[0]
+        nuc['pragma'] = SeqFeature(
+            FeatureLocation(start, stop),
+            strand=1,
+            type='##sequence-region pragma',
+            id=name,
+        )
+    for feat in record.features:
+        if feat.type == "region":
+            nuc['region'] = feat
```

Comment:
Is there documentation we could cite here to explain why we look for (or prefer) a `region` type? I only see this referenced once in [the GFF3 spec under the "circular genomes" section](https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md#circular-genomes). The rest of the examples in the GFF spec use the sequence region pragma.

Reply:
Addressed in [this thread](https://github.com/nextstrain/augur/pull/1351#discussion_r1418150947)...

Reply:
Thanks! But just to make sure I understand, did you find the `region` type when looking into circular genome annotations in the GFF spec or did it popup somewhere else outside of the spec?

Reply:
I've seen it elsewhere, but I can't remember where exactly. It's also part of the [SO](http://www.sequenceontology.org/browser/current_release/term/SO:0000001) list. (I've never seen source used outside of Nextstrain.) When working on these issues I was surprised that GFF and (especially) GenBank specs are hard to find / somewhat incomplete.